update for land IAU to ccpp-physics noahmp

tsga opened the pull request:
This ensures consistency between the NoahMP component model and the UFS WM/ccpp physics Noahmp after the land IAU capability is added to the later. Specifically the noahmpdrv_init function has three parameters (Land IAU related DDTs) passed as optional, enabling the component model to use the noahmpdrv_init function without any more changes right now.

Dependent on pull request: https://github.com/ufs-community/ccpp-physics/pull/222

Commit:
update for land IAU to ccpp-physics noahmp

tsga commented:
Tagging relevant people here. Please add (tag) anyone you think should be aware.
@uturuncoglu, @barlage, @CoryMartin-NOAA, @ClaraDraper-NOAA, @CatherineThomas-NOAA, @junwang-noaa

CoryMartin-NOAA reviewed (approved):
I don't know this code well but these changes seem correct from what I can tell

uturuncoglu commented on drivers/ccpp/noahmpdrv.F90:
```diff
@@ -100,9 +119,309 @@ subroutine noahmpdrv_init(lsm, lsm_noahmp, me, isot, ivegsrc, &
+!! \section arg_table_noahmpdrv_timestep_init Argument Table
+!! \htmlinclude noahmpdrv_timestep_init.html
+!!
+subroutine noahmpdrv_timestep_init (itime, fhour, delt, km,  ncols,         &      
```
@barlage @tsga I think this is a new routine. Does this need to be called by the component model? At this point, the component model is just calling init and run. What happens if we don't call from component. Maybe it is just called in the IAU case.

barlage replied:
I believe @tsga has this routine only doing IAU so it doesn't need to be called in the component model for now.

tsga replied:
> I believe @tsga has this routine only doing IAU so it doesn't need to be called in the component model for now.

Yes, this function (and noahmpdrv_finalize) is only called for IAU, so the component model just skips it.

uturuncoglu commented on drivers/ccpp/noahmpdrv.meta:
```diff
@@ -96,6 +96,230 @@
   dimensions = ()
```
As we discussed both files need to be pushed to NoahMP EMC repository too.

tsga replied:
> As we discussed both files need to be pushed to NoahMP EMC repository too.

Yes, they are now copied.

uturuncoglu reviewed (commented):
I have just put couple of comment to clarify the implementation.

tsga committed:
update for land iau

tsga committed:
Merge branch 'tmp' into feature/lnd_iau

* tmp:
  update for land iau

barlage commented:
@tsga thanks for adding the IAU code here; I had a comment on that yesterday that apparently I didn't send; have you tried to compile and maybe run the land component reg test?

@uturuncoglu I assume there will need to be additions [here](https://github.com/NOAA-EMC/noahmp/blob/3ac32f0db7a2a97d930f44fa5f060c983ff31ee8/CMakeLists.txt#L21)

tsga committed:
update CMakeLists

tsga commented:
> @tsga thanks for adding the IAU code here; I had a comment on that yesterday that apparently I didn't send; have you tried to compile and maybe run the land component reg test?
> 
> @uturuncoglu I assume there will need to be additions [here](https://github.com/NOAA-EMC/noahmp/blob/3ac32f0db7a2a97d930f44fa5f060c983ff31ee8/CMakeLists.txt#L21)

@barlage yes, the land component related reg tests passed after all the changes were made. I am now running the whole UWM regression tests.  

And I have now added the CMakeLists.txt you mentioned to @uturuncoglu to noahmp repo as well (It is also part of the UWM repo, so I had modified it in the UWM repo--probably why the tests didn't fail).

uturuncoglu commented:
@tsga Thanks. This looks fine. Just to be sure, did you run land component related RTs? If so, are they passing without any issue?

tsga commented:
> @tsga Thanks. This looks fine. Just to be sure, did you run land component related RTs? If so, are they passing without any issue?

@uturuncoglu yes, All of the regression tests listed in "rt.conf" pass. Yesterday, "datm_cdeps_lm4_gswp3_intel" did fail, but it passes now after I updated the LM4-driver repo. The new land iau test (control_c48_lnd_iau_intel) needs to be run with the -c switch because the baseline outputs are not yet copied in hera (and the other test platforms).

barlage commented:
@tsga what did you change in LM4? did you contact the LM4 group about those changes?

tsga commented:
> @tsga what did you change in LM4? did you contact the LM4 group about those changes?

No, I didn't change anything. I only needed to get my branch up to date with all the develop branch of UFSWM and its submodules, and everything works OK.

barlage commented:
> > @tsga what did you change in LM4? did you contact the LM4 group about those changes?
> 
> No, I didn't change anything. I only needed to get my branch up to date with all the develop branch of UFSWM and its submodules, and everything works OK.

OK, got it. I misinterpreted what you were saying.

ClaraDraper-NOAA reviewed (commented):
Looks OK to me - just one small comment. Thanks @tsga

ClaraDraper-NOAA commented:
@tsga I reviewed this repo as this is where the science is done. Would you like to review the related PRs too?

tsga commented:
> @tsga I reviewed this repo as this is where the science is done. Would you like to review the related PRs too?

Thank you @ClaraDraper-NOAA. I think this is enough. The other two repo's will be reviewed by the code owners for dependency etc.

tsga commented:
@uturuncoglu It seems the second test is failing because the files "lnd_iau_mod.F90" and "lnd_iau_mod.meta" are not copied to the noahmp directory. Can you please rerun the checks with the feature/lnd_iau branch contents and let me know if there is anything else to be done to merge this PR (the other related PRs are waiting for this to be complete).

uturuncoglu commented:
@tsga Okay. Run it again. Let's see what happens.

uturuncoglu commented on CMakeLists.txt:
```diff
@@ -19,6 +19,7 @@ list(APPEND _noahmp_cap_files drivers/nuopc/lnd_comp_kind.F90
 
 # CCPP interface
 list(APPEND _noahmp_ccpp_files drivers/ccpp/noahmpdrv.F90
+	                       noahmp/drivers/ccpp/lnd_iau_mod.F90
```
@tsga I think this is wrong. It must be `drivers/ccpp/lnd_iau_mod.F90`. The test still fails.

tsga replied:
Thank you @uturuncoglu. I just fixed it.

tsga committed:
Update CMakeLists.txt

fix file path

tsga commented:
@barlage Mike, now that the tests are all passed, can you please mark this PR as approved.

barlage reviewed (approved):
Confirmed four files are identical to ccpp-physics

tsga committed:
sync with uwm ccpp/physics

tsga committed:
minor edit to combine use lnd_iau_mod lines

FernandoAndrade-NOAA commented:
Testing for [2415](https://github.com/ufs-community/ufs-weather-model/pull/2415) has completed, please continue with the merge process, thank you.

jkbk2004 commented:
@uturuncoglu @barlage can you merge this pr?

uturuncoglu commented:
@jkbk2004 merged.
